### PR TITLE
Mnaboka/add snapshot

### PR DESCRIPTION
**Pull request by darkonie**

### CODE REVIEW

**Comment by darkonie** [comment]
@alberts @malnick @jfrazelle this is ready for initial review when you have some time.

File: api/snapshot.go [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+		}
+	}
+
+	// Null errors on every new run.
+	j.Errors = nil
+
+	t := time.Now()
+	j.LastSnapshotPath = fmt.Sprintf("%s/snapshot-%d-%02d-%02dT%02d:%02d:%02d-%d.zip", config.FlagSnapshotDir, t.Year(),
+		t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond())
+	j.Status = "Snapshot job started, archive will be available: " + j.LastSnapshotPath
```
Comment (malnick) [inline api/snapshot.go]: `archive will be available at:`?

File: api/snapshot.go [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+			j.Status = "Job failed"
+			errMsg := fmt.Sprintf("snapshot job timedout after: %s", time.Since(j.JobStarted))
+			j.Errors = append(j.Errors, errMsg)
+			log.Error(errMsg)
+			j.cancelChan <- true
+			return
+		}
+	}(jobIsDone, j)
+
+	// makesure we always cancel a timeout goroutine when the report is finished.
```
Comment (malnick) [inline api/snapshot.go]: makesure -> make sure

File: api/snapshot.go [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+	defer func(jobIsDone chan bool) {
```
Comment (malnick) [inline api/snapshot.go]: 👍

File: api/snapshot.go [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+	}(jobIsDone)
+
+	// Update job running field.
+	j.start()
+	defer j.stop()
+
+	// create a zip file
+	zipfile, err := os.Create(j.LastSnapshotPath)
+	if err != nil {
+		j.Status = "Job failed"
```
Comment (malnick) [inline api/snapshot.go]: It seems like `.Status` and `.Errors` both provide stringified status messages but are somewhat duplicated results. You could make `.Status` a bool and only be verbose with `.Errors` when `.Status` is false (I guess status could be `.Success` in this context).

File: api/snapshot.go [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+
+	zipWriter := zip.NewWriter(zipfile)
+	defer zipWriter.Close()
+
+	// place a summaryErrorsReport.txt in a zip archive which should provide info what failed during the logs collection.
+	summaryErrorsReport := new(bytes.Buffer)
+	defer func(zipWriter *zip.Writer, summaryReport *bytes.Buffer) {
+		zipFile, err := zipWriter.Create("summaryErrorsReport.txt")
+		if err != nil {
+			j.Status = "Could not append a summaryErrorsReport.txt to a zip file, node"
```
Comment (malnick) [inline api/snapshot.go]: See comment above, I feel like we duplicate the meaning of `.Errors` here

File: api/snapshot.go [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+
+	node, _, ok, err := j.isSnapshotAvailable(snapshotName, config, DCOSTools)
+	if err != nil {
+		return prepareResponseWithErr(http.StatusServiceUnavailable, err)
+	}
+	if ok {
+		url := fmt.Sprintf("http://%s:%d%s/report/snapshot/delete/%s", node, config.FlagPort, BaseRoute, snapshotName)
+		j.Status = "Attempting to delete a snapshot on a remote host. POST " + url
+		log.Debug(j.Status)
+		timeout := time.Duration(time.Second * 5)
```
Comment (malnick) [inline api/snapshot.go]: Did we want a configurable timeout here? This isn't a blocker, but I think eventually the UI might need to configure this on the fly.

File: api/snapshot.go [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+			continue
+		}
+		if resp.Header.Get("Content-Encoding") == "gzip" {
+			fileName += ".gz"
+		}
+
+		// put all logs in a `ip_role` folder
+		zipFile, err := zipWriter.Create(filepath.Join(node.IP+"_"+node.Role, fileName))
+		if err != nil {
+			resp.Body.Close()
```
Comment (malnick) [inline api/snapshot.go]: `resp.Body.Close()` should drop in on line 398 as `defer resp.Body.Close()`. Then you can remove it here and below the if on 414.

**Comment by malnick** [comment]
This is looking solid. I had a few comments, nothing to big. The one thing that could be a nice refactor downstream of this PR is breaking snapshot.go into smaller pieces (files). Right now it's getting pretty big and hard to follow. You could make a new pkg potentially, or without overcomplicating it, could make a few broken down files with `snapshot_*` prefix. That's just a nit, not a blocker. 

Should definitely check out the comment I had on the `resp.Body.Close()`, it appears you could call it right after the err handling with `defer` but instead called it twice explicitly instead after the `resp` is made.

File: api/snapshot.go (hunk 7 above)
Comment (darkonie) [inline api/snapshot.go]: i'd avoid calling defer in a loop

File: api/snapshot.go (hunk 6 above)
Comment (darkonie) [inline api/snapshot.go]: done

File: api/snapshot.go (hunk 5 above)
Comment (darkonie) [inline api/snapshot.go]: Status and Errors are a bit different, Status provides a live update what is the status of a job, when Errors is a collection of errors occurred during the job running

Comment (darkonie) [inline api/snapshot.go]: Status and Errors are a bit different, Status provides a live update what is the status of a job, when Errors is a collection of errors occurred during the job running

File: api/snapshot.go (hunk 1 above)
Comment (darkonie) [inline api/snapshot.go]: done

File: api/snapshot.go (hunk 2 above)
Comment (darkonie) [inline api/snapshot.go]: done
